Use Preferences for using custom hdf5 library

JoshuaLampert opened the pull request:
This switches to Preferences.jl for using a system-provided hdf5 library and replaces the additional build step. ~~Additionally, I allowed additional names of the library binaries as in my case, e.g., the binaries are located in `/usr/lib/x86_64-linux-gnu/libhdf5_openmpi.so`  and `/usr/lib/x86_64-linux-gnu/libhdf5_openmpi_hl.so`.~~

closes #1037 

cc @ranocha, @sloede

JoshuaLampert commented:
Honestly, I don't understand why the tests complain about `Package HDF5 does not have HDF5_JLL in its dependencies`. Isn't HDF5_jll listed as a dependency [here](https://github.com/JuliaIO/HDF5.jl/blob/master/Project.toml#L7) or am I missing something stupid?

sloede commented on Project.toml:
```diff
@@ -2,21 +2,22 @@ name = "HDF5"
 version = "0.16.14"
 
+[compat]
+Compat = "3.1.0, 4"
+HDF5_jll = "~1.10.5, ~1.12.0"
+Requires = "1.0"
+julia = "1.3"
```
```suggestion
Compat = "3.1.0, 4"
HDF5_jll = "~1.10.5, ~1.12.0"
Preferences = "1.3"
Requires = "1.0"
julia = "1.3"
```
Add compat for Preferences?

sloede commented on src/api/api.jl:
```diff
@@ -1,16 +1,46 @@
+
+const libpath = @load_preference("libhdf5path", nothing)
+if libpath === nothing
+    using HDF5_JLL
```
```suggestion
    using HDF5_jll
```
Should fix the current CI issue

JoshuaLampert replied:
Ok, I really missed something stupid :smile:. Thanks and done.

sloede commented on src/api/api.jl:
```diff
@@ -1,16 +1,46 @@
 module API
 
-import Libdl
+using Libdl
```
```suggestion
using Libdl: dlopen, dlclose, dlpath, RTLD_LAZY, RTLD_NODELETE
```
To be consistent with the other `using` statements?

sloede commented on docs/src/index.md:
````diff
@@ -27,24 +27,35 @@ Starting from Julia 1.3, the HDF5 binaries are by default downloaded using the `
 
+to set the MPI preferences, see the [documentation of MPI.jl](https://juliaparallel.org/MPI.jl/stable/configuration/) and then:
+
+```julia
````
Maybe set the preferences first and then use MPI's system binary? Or will that cause issues? I'd also add a (short) explanation when and why the MPI library change is necessary (and when it isn't, i.e., when switching to another, non-MPI enabled HF5 library or (possibly?) when not using MPI at all).

JoshuaLampert replied:
I thought setting the MPIPreference first is more intuitive since this setting is more basic and HDF5 comes later as it is more specific. But, you can do it both ways...
I will add a short explanation as you suggested.

sloede commented on docs/src/mpi.md:
````diff
@@ -34,9 +34,16 @@ for details.
+julia> set_preferences!(
+           UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
+           "libhdf5path" => "/path/to/your/hdf5/binaries/", force = true)
+```
````
Also mention here that the underlying MPI library might have to be changed (using MPIPreferences.jl)?

sloede replied:
Also, here and in the section on custom binaries you should probably include the command to use if you want to reset the usage to the HDF5_jll provided library.

JoshuaLampert replied:
Regarding your first point: How to set the MPI library with MPIPreferences.jl is described in the previous section, see [here](https://juliaio.github.io/HDF5.jl/stable/mpi/#.-Using-system-provided-MPI-libraries).
Regarding your second point: You mean deleting the LocalPreferences.toml or `@delete_preferences!` or something else?

sloede replied:
> How to set the MPI library with MPIPreferences.jl is described in the previous section

Gotcha 👍



> You mean deleting the LocalPreferences.toml or `@delete_preferences!` or something else?

Yes. Many Julia users are not aware (yet) of how the Preferences mechanism works. For them, it might be surprising to not be able to load HDF5 anymore and no helpful error message on how to fix it. Maybe explain it here in the docs, but also consider adding it to the error messages below.

JoshuaLampert replied:
Ok, I will do that.

JoshuaLampert replied:
I extended the documentation with your proposals in f292a51d8404beb52d7bf62d5ece05c9414a6b43.

sloede replied:
Great, looks good!

sloede commented on src/api/api.jl:
```diff
@@ -1,16 +1,46 @@
+    libhdf5 = find_library_alt(["libhdf5", "libhdf5_openmpi", "libhdf5_mpich"], libpaths; flags=flags)
+    libhdf5_hl = find_library_alt(["libhdf5_hl", "libhdf5_openmpi_hl", "libhdf5_mpich_hl"], libpaths; flags=flags)
+
+    isempty(libhdf5) && error("libhdf5 could not be found")
+    isempty(libhdf5_hl) && error("libhdf5_hl could not be found")
```
This will prevent the user from loading HDF5 if the custom library has been (re)moved since it `libhdf5path` was set. IMHO (and from past experience) it would be good to add a note to the docs on how to proceed in that case.

mkitti commented:
Please run the formatter. See the contrib folder for a utility.

Also this is breaking so I think we will want to version bump to 0.17 unless we implement a fallback. I'm also wondering if we could include a migration utility.

sloede commented:
> Also this is breaking so I think we will want to version bump to 0.17 unless we implement a fallback. I'm also wondering if we could include a migration utility.

MPI.jl had performed a similar migration last year, from configuration via environment variables to using Preferences. Back then they solved this by adding a [migration guide to the docs](https://juliaparallel.org/MPI.jl/stable/configuration/#Migration-from-MPI.jl-v0.19-or-earlier) and included [notes for HPC administrators](https://juliaparallel.org/MPI.jl/stable/configuration/#Notes-to-HPC-cluster-administrators) on how to configure this on clusters with managed MPI/HDF5 installations. Maybe this can serve as a blueprint for this PR here as well?

ranocha commented:
> Also this is breaking so I think we will want to version bump to 0.17 unless we implement a fallback. I'm also wondering if we could include a migration utility.

This should probably be mentioned in https://github.com/JuliaIO/HDF5.jl/blob/master/HISTORY.md, too, I guess?

mkitti commented:
Yes, please document this as much as possible. If we do provide migration instructions, that should also be tested.

sloede commented:
> I provided some migration instructions.

This reads already very well! One suggestion from my side: Maybe you can give an example for admins of clusters what the Preferences file could look like and maybe explicitly mention that for backward compatibility it is recommended to both set the environment variables and use the Preferences file for the time being. But these are just suggestions, they are not critical IMHO. Other than that, this PR LGTM!

JoshuaLampert commented:
I've also added an example of how the LocalPreferences.toml could look like and added a recommendation to also set the environment variable for backward compatibility.

sloede commented:
> I've also added an example of how the LocalPreferences.toml could look like and added a recommendation to also set the environment variable for backward compatibility.

Great, thanks!

ranocha commented:
I think this PR is really nice! You could consider adding a CI job for a system libhdf5 with MPI enabled also on Julia v1 to test this setup on the latest stable release.

JoshuaLampert commented:
Ok, I'll try. I've also noticed that the current system libhdf5 + mpich CI job still uses environment variables instead of MPIPreferences to use the system MPI. I'll also adapt that, right?

JoshuaLampert commented:
I'm sorry, but I don't know why [this error](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) comes up. I tried to use a similar setup than we have in [P4est.jl](https://github.com/trixi-framework/P4est.jl/blob/main/.github/workflows/ci.yml#L118), but somehow it doesn't recognize the configure_packages.jl file. I would be thankful for a hint!

JoshuaLampert commented on .github/workflows/CI.yml:
```diff
@@ -27,7 +28,17 @@ jobs:
           sudo apt-get install mpich libhdf5-mpich-dev
           echo "JULIA_HDF5_PATH=/usr/lib/x86_64-linux-gnu/hdf5/mpich/" >> $GITHUB_ENV
-          echo "JULIA_MPI_BINARY=system" >> $GITHUB_ENV
+      - uses: julia-actions/setup-julia@v1
+        with:
+          version: ${{ matrix.version }}
+          arch: ${{ matrix.arch }}
+          show-versioninfo: true
+      - uses: julia-actions/cache@v1
+      - name: Configure MPI.jl
+        shell: julia --color=yes {0}
+        run: |
+          @show pwd()
+          include(joinpath(pwd(), "test", "configure_packages.jl"))
```
I am not sure if I understand correctly (I am not very experienced with setting up CI workflows), but do you mean that the "Configure MPI.jl " step should be **after** running `actions/checkout`?

ranocha replied:
Yes. `actions/checkout` should clone the git repo of HDF5.jl. Without running it, you don't have access to its files.

ranocha commented on .github/workflows/CI.yml:
```diff
@@ -27,8 +28,18 @@ jobs:
           echo "JULIA_HDF5_PATH=/usr/lib/x86_64-linux-gnu/hdf5/mpich/" >> $GITHUB_ENV
-          echo "JULIA_MPI_BINARY=system" >> $GITHUB_ENV
       - uses: actions/checkout@v3
+      - uses: julia-actions/setup-julia@v1
+        with:
+          version: ${{ matrix.version }}
+          arch: ${{ matrix.arch }}
+          show-versioninfo: true
```
There is already a `julia-actions/setup-julia` job below. Please use it only once

JoshuaLampert replied:
Done.

JoshuaLampert commented on .github/workflows/CI.yml:
```diff
@@ -29,12 +29,6 @@ jobs:
-          arch: ${{ matrix.arch }}
-          show-versioninfo: true
-      - uses: julia-actions/cache@v1
       - name: Configure MPI.jl
```
I think I got it now. Thank you for your guidance!

JoshuaLampert commented:
I think the failing test is not related to this PR. It's also failing on master with parallel HDF5 enabled.

simonbyrne commented on src/api/api.jl:
```diff
@@ -1,16 +1,53 @@
+    flags = RTLD_LAZY | RTLD_NODELETE  # RTLD_NODELETE may be needed to avoid segfault (#929)
+
+    libhdf5 = find_library_alt(
+        ["libhdf5", "libhdf5_openmpi", "libhdf5_mpich"], libpaths; flags=flags
```
Are these standard names? would it be better to just provide separate prefs for each library?

JoshuaLampert replied:
Ok, I will do that :+1: .

JoshuaLampert commented:
As with the new HDF5_jll release MPI is supported, I updated the docs accordingly. So, this is ready to go from my side. What do you think, @mkitti? CI failures seem unrelated to this PR (?)

ranocha commented:
The CI failures on julia-nightly are `caused by: UndefVarError: `_memcpy!` not defined`. That's not related to this PR as far as I can tell :+1: 

The error
```
virtual dataset: Error During Test at /home/runner/work/HDF5.jl/HDF5.jl/test/virtual_dataset.jl:31
  Test threw exception
  Expression: size(d) == (3, 2)
  HDF5.API.H5Error: Error getting dataspace
  libhdf5 Stacktrace:
    [1] H5FO_dest: Object cache/Unable to release object
        objects still in open object info set
     ⋮
```
in https://github.com/JuliaIO/HDF5.jl/actions/runs/5112753923/jobs/9191167908?pr=1061 seems to be real, though? Can you reproduce it locally? Has anyone seen it before?

JoshuaLampert commented:
> Can you reproduce it locally?
Yes, I can reproduce this error locally, when I use a system provided HDF5 library (v1.10.6), where I get even another error message. What confuses me is that julia v1.6 seems to work, but not v1.9.

JoshuaLampert commented:
Ah, I see. On julia v1.9 libhdf5 v1.10.4 is used, while on julia v.1.6 libhdf5 v.1.14.0. Probably that's the difference that makes v1.9 fail and v1.6 pass.

ranocha commented:
Can you use the newer libhdf5 also with Julia v1.9?

JoshuaLampert commented:
Why do they actually use different libhdf5 versions, in the first place? Both run on ubuntu 20.04.

ranocha commented:
That's weird - the Julia v1.6 run reports the expected version 1.10.4 when installing the system packages, e.g.,
https://github.com/JuliaIO/HDF5.jl/actions/runs/5112753923/jobs/9191167775#step:2:120
However, it shows 1.14.0 when running the tests in
https://github.com/JuliaIO/HDF5.jl/actions/runs/5112753923/jobs/9191167775#step:7:171
That's exactly the version from HDF5_jll.jl. Does it not pick the correct system library?

JoshuaLampert commented:
Yes, that's strange. Locally, with julia v1.6.7 it picks up the correct local libhdf5 library for me.

ranocha commented:
If the problem is an old libhdf5 library, would moving to a newer version of the CI image help?

JoshuaLampert commented:
You mean using ubuntu-latest instead of ubuntu-20.04?

ranocha commented:
Something like that

ranocha commented:
I vaguely remember that there were some issues with `Pkg.test` not picking up preferences correctly, e.g., https://github.com/JuliaLang/Pkg.jl/issues/2500 and https://github.com/JuliaLang/Pkg.jl/issues/3389. One comment in https://github.com/JuliaLang/Pkg.jl/issues/2500 seems to suggest that the issue may be fixed in Julia v1.7 - but not in Julia v1.6? If so, this would explain why the preferences are not picked up correctly?

JoshuaLampert commented:
That's exactly what I tried in 338791c. Didn't work, see https://github.com/JuliaIO/HDF5.jl/actions/runs/5114015916/jobs/9193838265. However, there it was another error...

JoshuaLampert commented:
> I vaguely remember that there were some issues with Pkg.test not picking up preferences correctly, e.g., https://github.com/JuliaLang/Pkg.jl/issues/2500 and https://github.com/JuliaLang/Pkg.jl/issues/3389. One comment in https://github.com/JuliaLang/Pkg.jl/issues/2500 seems to suggest that the issue may be fixed in Julia v1.7 - but not in Julia v1.6? If so, this would explain why the preferences are not picked up correctly?

Yes, you're right. Local tests confirm that `Pkg.test` picks the right preferences with julia v1.8.0, but not with v1.6.7 and also not for julia v1.7.3 (in my previous comment I didn't look whether the preferences are picked up correctly with `Pkg.test`, but only if the right libhdf5 library is picked up in `HDF5.API.h5_get_libversion()`). Should I change the julia version for the system libhdf5 + mpich tests to v1.8 or should we just drop the version and only keep v1.9?

ranocha commented:
Could you please try a workaround similar to https://github.com/maleadt/LLVM.jl/commit/ebb44ee1acbde749f3cfb96e51e880186cd810de for Julia v1.6 here?

JoshuaLampert commented:
> Could you please try a workaround similar to https://github.com/maleadt/LLVM.jl/commit/ebb44ee1acbde749f3cfb96e51e880186cd810de for Julia v1.6 here?

I did that. Now also julia v1.6 recognizes the local libhdf5 installation correctly and there is only the problem with the older libhdf5 library.

ranocha commented:
Great, thanks! I think it would be good to get feedback/help from the HDF5.jl maintainers at this point to fix the CI problem.

mkitti commented:
I'm a bit confused why the one test is failing now. Could someone summarize the current understanding of why it fails?

JoshuaLampert commented:
For me, it's also failing on master with the same error using a local libhdf5 v1.10.6 on

```julia
julia> versioninfo()
Julia Version 1.9.0
Commit 8e630552924 (2023-05-07 11:25 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 1 on 8 virtual cores
Environment:
  JULIA_HDF5_PATH = /usr/lib/x86_64-linux-gnu/hdf5/openmpi
```
I'm rather confused why it's **not** failing in the other CI tests.

simonbyrne commented:
I think the problem is that `h5d_get_space` is broken for virtual datasets for libhdf5 1.10. At the moment, we only test custom hdf5 build on Julia 1.3, and where [we disable the virtual dataset tests](https://github.com/JuliaIO/HDF5.jl/blob/cfba7062d21761b7d9700a55e9eb632d01a3a643/test/runtests.jl#L67-L70). 

We previously [thought it was a buggy JLL](https://github.com/JuliaIO/HDF5.jl/pull/1019#discussion_r997691382), but it could just be all v1.10. I was able to replicate the error with a local build using libhdf5 1.10.7.

I suggest we just disable the virtual dataset tests if using libhdf5 <1.12.

JoshuaLampert commented:
That totally makes sense. Thanks, @simonbyrne! I disabled the virtual datasets tests for libhdf5 < 1.12.

JoshuaLampert commented:
Loading MPI.jl _before_ HDF5.jl is not required anymore with Preferences.jl, right? I changed the docs accordingly.

ranocha commented:
Status of this PR is that HDF5.jl still uses Requires.jl to load the MPI code, doesn't it? If so, it's still necessary to first load MPI.jl as far as I know

ranocha reviewed (approved):
Thanks! Looks good to me. Let's keep fingers crossed that tests pass 🤞

mkitti commented:
Could we issue a warning if we detect that the environment variable `JULIA_HDF5_PATH` is set?

mkitti commented on test/configure_packages.jl:
```diff
@@ -0,0 +1,26 @@
+import UUIDs, Preferences
+Preferences.set_preferences!(
+    UUIDs.UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
+    "libhdf5" => joinpath(JULIA_HDF5_PATH, "libhdf5.so");
```
Could libhdf5.so also be in a `lib` or `lib64` subdirectory?

JoshuaLampert replied:
Not as it is now. Primarily, I created this file for CI, where we know, where the binaries are located. Now, that we suggest the user to have a look at this file, it would probably be nice to also allow the binaries to be in `joinpath(JULIA_HDF5_PATH, "lib")` or `joinpath(JULIA_HDF5_PATH, "lib64")`.

mkitti commented on test/configure_packages.jl:
```diff
@@ -0,0 +1,26 @@
+import UUIDs, Preferences
+Preferences.set_preferences!(
+    UUIDs.UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
+    "libhdf5" => joinpath(JULIA_HDF5_PATH, "libhdf5.so");
```
```suggestion
    "libhdf5" => joinpath(JULIA_HDF5_PATH, "libhdf5." * Libdl.dlext);
```

mkitti commented on test/configure_packages.jl:
```diff
@@ -0,0 +1,26 @@
+)
+Preferences.set_preferences!(
+    UUIDs.UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
+    "libhdf5_hl" => joinpath(JULIA_HDF5_PATH, "libhdf5_hl.so");
```
```suggestion
    "libhdf5_hl" => joinpath(JULIA_HDF5_PATH, "libhdf5_hl." * Libdl.dlext);
```

mkitti commented on test/configure_packages.jl:
```diff
@@ -0,0 +1,26 @@
+using Pkg
```
```suggestion
using Pkg, Libdl
```

mkitti commented on docs/src/index.md:
````diff
@@ -27,24 +27,58 @@ Starting from Julia 1.3, the HDF5 binaries are by default downloaded using the `
+    UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
+    "libhdf5_hl" => "/usr/lib/x86_64-linux-gnu/hdf5/mpich/libhdf5_hl.so", force = true)
+```
+
````
```suggestion

```suggestion

Also see the file `test/configure_packages.jl` for an example.

```

```

mkitti reviewed (changes requested):
The warning should be done during `__init__` since the environment can change after precompilation occurs. Also, since we are encouraging people to set both the enviornment variable and Preferences.jl we should not warn if they are in fact using both.

I'm also wondering if we should just add a convenience method `HDF5.API.set_libraries!`.

```julia
const HDF5_JL_UUID = UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f")

function set_libraries!(libhdf5 = nothing, libhdf5_hl = nothing; force = true)
   if isnothing(libhdf5) && isnothing(libhdf5_hl)
      delete_preferences!(HDF5_JL_UUID, "libhdf5")
      delete_preferences!(HDF5_JL_UUID, "libhdf5_hl")
      delete_preferences!("HDF5_jll", "libhdf5", "libhdf5_hl")
      @info "The libraries from HDF5_jll will be used."
    elseif isnothing(libhdf5) && isnothing(libhdf5_hl)
        throw(ArgumentError("Specify either no positional arguments or both positional arguments."))
   else
       isfile(libhdf5) || throw(ArgumentError("$libhdf5 is not a file that exists.")
       isfile(libhdf5_hl) || throw(ArgumentError("$libhdf5_hl is not a file that exists.")
       set_preferences!(HDF5_JL_UUID, "libhdf5" => libhdf5; force)
       set_preferences!(HDF5_JL_UUID, "libhdf5_hl" => libhdf5_hl; force)
       # Also set the HDF5_jll override settings in case some other package tries to use HDF5_jll
       set_preferences!(
           "HDF5_jll",
           "libhdf5_path" => libhdf5,
           "libhdf5_jl_path" => libhdf5_jl
       )
   end
   @info "Please restart Julia and reload HDF5.jl for the library changes to take effect"
end
```

See https://github.com/jheinen/GR.jl/blob/master/src/preferences.jl for further inspiration. Perhaps that is for another pull request.

mkitti commented:
I think we should also configure document configuring HDF5_jll as well in case some other package tries to load that.

See https://docs.binarybuilder.org/stable/jll/#Overriding-specific-products

simonbyrne commented on docs/src/mpi.md:
````diff
@@ -31,12 +31,29 @@ See the [MPI.jl
+```julia
+julia> using Preferences, UUIDs
+
+julia> set_preferences!(
+           UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
+           "libhdf5" => "/path/to/your/libhdf5.so", force = true)
+julia> set_preferences!(
+           UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
+           "libhdf5_hl" => "/path/to/your/libhdf5_hl.so", force = true)
````
You should be able to set multiple properties:
```suggestion
julia> set_preferences!(
           UUID("f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"), # UUID of HDF5.jl
           "libhdf5" => "/path/to/your/libhdf5.so",
           "libhdf5_hl" => "/path/to/your/libhdf5_hl.so",
           force = true)
```

mkitti commented:
If you would prefer, I would be fine merging this as is and implementing the above later.

JoshuaLampert commented:
I addressed most of the suggestions. The remaining parts can be done in a separate PR.

sloede commented:
🥳
